src/trackinfoobject.cpp

```diff
@@ -184,26 +178,37 @@ namespace {
 }
 
 void TrackInfoObject::setMetadata(const Mixxx::TrackMetadata& trackMetadata) {
-    // TODO(XXX): This involves locking the mutex for every setXXX
-    // method. We should figure out an optimization where there are private
-    // setters that don't lock the mutex.
-    setArtist(trackMetadata.getArtist());
-    setTitle(trackMetadata.getTitle());
-    setAlbum(trackMetadata.getAlbum());
-    setAlbumArtist(trackMetadata.getAlbumArtist());
-    setYear(trackMetadata.getYear());
-    setGenre(trackMetadata.getGenre());
-    setComposer(trackMetadata.getComposer());
-    setGrouping(trackMetadata.getGrouping());
-    setComment(trackMetadata.getComment());
-    setTrackNumber(trackMetadata.getTrackNumber());
-    setChannels(trackMetadata.getChannels());
-    setSampleRate(trackMetadata.getSampleRate());
-    setDuration(trackMetadata.getDuration());
-    setBitrate(trackMetadata.getBitrate());
-
-    if (trackMetadata.isReplayGainValid()) {
-        setReplayGain(trackMetadata.getReplayGain());
+    {
+        // enter locking scope
+        QMutexLocker lock(&m_qMutex);
+
+        bool modified = false;
+        modified |= modifyFieldValue(&m_sArtist, trackMetadata.getArtist());
+        modified |= modifyFieldValue(&m_sTitle, trackMetadata.getTitle());
+        modified |= modifyFieldValue(&m_sAlbum, trackMetadata.getAlbum());
+        modified |= modifyFieldValue(&m_sAlbumArtist, trackMetadata.getAlbumArtist());
+        modified |= modifyFieldValue(&m_sYear, trackMetadata.getYear());
+        modified |= modifyFieldValue(&m_sGenre, trackMetadata.getGenre());
+        modified |= modifyFieldValue(&m_sComposer, trackMetadata.getComposer());
+        modified |= modifyFieldValue(&m_sGrouping, trackMetadata.getGrouping());
+        modified |= modifyFieldValue(&m_sComment, trackMetadata.getComment());
+        modified |= modifyFieldValue(&m_sTrackNumber, trackMetadata.getTrackNumber());
+        modified |= modifyFieldValue(&m_iChannels, trackMetadata.getChannels());
+        modified |= modifyFieldValue(&m_iSampleRate, trackMetadata.getSampleRate());
+        modified |= modifyFieldValue(&m_iDuration, trackMetadata.getDuration());
+        modified |= modifyFieldValue(&m_iBitrate, trackMetadata.getBitrate());
+
+        const bool replayGainUpdated =
+                modifyFieldValue(&m_fReplayGain, trackMetadata.getReplayGain());
+        modified |= replayGainUpdated;
+
+        if (modified) {
+            setDirtyAndUnlock(&lock, true);
```

Comment:
This setDirtyAndUnlock improves the locking a lot but it is a little bit confusion, because the function name hides some details and the lock is sometimes released inside the scope. 

What about extend the locker with the dirty emit management. (Not sure if it is an improvement, just an idea) 
```
{ // scope 
DirtyFlagLocker dirtyFlagLocker (&m_qMutex);

// ... property changes 

dirtyFlagLocker.setDirty(true) 
}
if (replayGainUpdated) {
...